### ISSUE TITLE
Prediction Signal?

### ISSUE THREAD

**Issue by stabilus** [teddykoker/blog#6]
I admit your project and math are over my head. I'm still struggling to understand your position function and with that I'm trying to come up with a simple

- input format
- that yields a simple output format (that I can interprete as a signal to buy, sell or hold)

How would that be done in your code?

**Comment by teddykoker** [comment]
Which strategy are you referring to?

**Comment by stabilus** [comment]
I'm talking about 2019-06-04-trading-with-reinforcement-learning-in-python-part-ii

**Comment by teddykoker** [comment]
Look at `positions()`:

```python
def positions(x, theta):
    M = len(theta) - 2
    T = len(x)
    Ft = np.zeros(T)
    for t in range(M, T):
        xt = np.concatenate([[1], x[t - M:t], [Ft[t - 1]]])
        Ft[t] = np.tanh(np.dot(theta, xt))
    return Ft
```
`x` is a standardized series of prices; `theta` are the parameters found via gradient ascent. The output of the function will be an array the same length of input `x` with each value falling between 1 and -1. Values above 0 would indicate buy/hold, below 0 would indicate sell/short. 

To standardize input data and find parameters `theta`, please read the rest of the post.

**Comment by stabilus** [comment]
Thanks for your clarification. However, a lot of questions remain:


  *   Tanh returns values in the range -1 to 1. In fact most Ft values created are either -1 or 1, with only a very small portion being in between. So, the Ft value returned could be rescaled to match a scale 0 to 1 and I assume that value could then represent the portion of the portfolio that should be invested in the asset. i.e. 0 = sell all, 1 buy all, 0.5= invest 50%. Correct?
  *   x in the positions() function is fed x_train which is a representation of tick to tick price differences expressed as a multiple of standard deviations. While I can reformat my price values accordingly to get predictions, I don’t exactly understand  the purpose of this reformatting of the price differences.
  *   How can it be correct that the “buy and hold” strategy yields near 0 throughout the entire duration of the recorded prices (2013 til now) when the price went up 1800x? How am I to understand the buy and hold strategy returns?
  *   Now, let’s assume, I would like to add more features to the training, say volume per tick and the 1 minute RSI. I end up with a multivariate environment. How could that be integrated in your code?
  *   GPU training: As the model gets more complex, training time becomes an issue. Is there a way to model all of this in a keras neural network and have keras take care of the math and the handling of the GPU which would probably significantly reduce training time?
  *   How does the positions function handle spreads between buy / sell prices? Market buys pay the higher sell price, market sells receive the lower buy price. The spread at an exchange like bitstamp varies and can be quite significant. Frequent buys and sells are bound to lose money on the spread.

There are more open issues about this code, but let’s see whether you’d be willing to clarify these. Thanks

Urs

**Comment by teddykoker** [comment]
1. My mistake, I have edited the original reply.
2. This [standardization](https://en.wikipedia.org/wiki/Feature_scaling#Standardization_(Z-score_Normalization)) is commonly done in machine learning to improve gradient descent. When standardizing your data, it is important to use the same mean and std. you trained with.
3. The charts in the post are a series of ticks, meaning the whole chart is a matter of minutes.
4. You can concatenate more features to the input vector and add respective parameters to theta.
5. It is fairly straightforward to recreate the model using an ML library. I have done so with PyTorch; it is just a RNN with a custom loss function using the Sharpe ratio.
6. This posts was manly done for educational purposes, and it will in no way be an actually profitable strategy without major modifications. It does make many assumptions about slippage and liquidity, which obviously need to be addressed when testing a more legitimate strategy.
